pkg/controller/servicetopology/servicetopology.go

```diff
@@ -0,0 +1,304 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/openyurtio/openyurt/pkg/controller/servicetopology/adapter"
+	"github.com/openyurtio/openyurt/pkg/yurthub/filter/servicetopology"
+	nodepoolv1alpha1 "github.com/openyurtio/yurt-app-manager-api/pkg/yurtappmanager/apis/apps/v1alpha1"
+	yurtinformers "github.com/openyurtio/yurt-app-manager-api/pkg/yurtappmanager/client/informers/externalversions"
+)
+
+const (
+	endpointsWorkerSize     = 2
+	endpointsliceWorkerSize = 4
+)
+
+type ServiceTopologyController struct {
+	endpointsQueue         workqueue.RateLimitingInterface
+	endpointsliceQueue     workqueue.RateLimitingInterface
+	serviceInformerSynced  cache.InformerSynced
+	nodepoolInformerSynced cache.InformerSynced
+	endpointsAdapter       adapter.Adapter
+	endpointsliceAdapter   adapter.Adapter
+	svcLister              corelisters.ServiceLister
+}
+
+func NewServiceTopologyController(client kubernetes.Interface,
+	sharedInformers informers.SharedInformerFactory,
+	yurtInformers yurtinformers.SharedInformerFactory) (*ServiceTopologyController, error) {
+	epSliceAdapter, err := getEndpointSliceAdapter(client, sharedInformers)
+	if err != nil {
+		return nil, fmt.Errorf("get endpointslice adapter with error: %v", err)
+	}
+	endpointsLister := sharedInformers.Core().V1().Endpoints().Lister()
+	epAdapter := adapter.NewEndpointsAdapter(client, endpointsLister)
+
+	epQueue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	epSliceQueue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+
+	serviceInformer := sharedInformers.Core().V1().Services()
+	nodepoolInformer := yurtInformers.Apps().V1alpha1().NodePools()
+	sc := &ServiceTopologyController{
+		endpointsQueue:         epQueue,
+		endpointsliceQueue:     epSliceQueue,
+		serviceInformerSynced:  serviceInformer.Informer().HasSynced,
+		nodepoolInformerSynced: nodepoolInformer.Informer().HasSynced,
+		svcLister:              serviceInformer.Lister(),
+		endpointsAdapter:       epAdapter,
+		endpointsliceAdapter:   epSliceAdapter,
+	}
+	serviceInformer.Informer().AddEventHandler(sc.getServiceHandler())
+	nodepoolInformer.Informer().AddEventHandler(sc.getNodepoolHandler())
+	return sc, nil
+}
+
+func (s *ServiceTopologyController) Run(stopCh <-chan struct{}) {
+	defer runtime.HandleCrash()
+	defer s.endpointsQueue.ShutDown()
+	defer s.endpointsliceQueue.ShutDown()
+
+	klog.Info("starting the service topology controller")
+	defer klog.Info("stopping the service topology controller")
+	if !cache.WaitForCacheSync(stopCh, s.serviceInformerSynced, s.nodepoolInformerSynced) {
+		klog.Error("sync service topology controller timeout")
+		return
+	}
```

Comment:
how about add a log message here after cache synced?

Reply:
because nodePool can not be synced before `yurt-app-manager` is ready.

Reply:
ok